aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,319 @@
+}
+
+impl TryFrom<u64> for HashToStructureSuite {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
+            0x0001000000000000 => Ok(HashToStructureSuite::Bls12381g1XmdSha256SswuRo),
+            0x0002000000000000 => Ok(HashToStructureSuite::Bls12381g2XmdSha256SswuRo),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Tid, Default)]
+pub struct AlgebraContext {
+    objs: Vec<Rc<dyn Any>>,
+}
+
+impl AlgebraContext {
+    pub fn new() -> Self {
+        Self { objs: Vec::new() }
+    }
+}
+
+macro_rules! structure_from_ty_arg {
+    ($context:expr, $typ:expr) => {{
+        let type_tag = $context.type_to_type_tag($typ).unwrap();
+        Structure::from_type_tag(&type_tag)
+    }};
+}
+
+macro_rules! store_obj {
+    ($context:expr, $obj:expr) => {{
+        let target_vec = &mut $context.extensions_mut().get_mut::<AlgebraContext>().objs;
+        let ret = target_vec.len();
+        target_vec.push(Rc::new($obj));
+        ret
+    }};
+}
+
+macro_rules! abort_unless_feature_enabled {
+    ($context:ident, $feature:expr) => {
+        if !$context.get_feature_flags().is_enabled($feature) {
+            return Err(SafeNativeError::Abort {
+                abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+            });
+        }
+    };
+}
+
+macro_rules! get_obj_pointer {
+    ($context:expr, $handle:expr) => {{
+        $context.extensions_mut().get_mut::<AlgebraContext>().objs[$handle].clone()
+    }};
+}
+
+macro_rules! ark_serialize_internal {
+    (
+        $gas_params:expr,
+        $context:expr,
+        $structure:expr,
+        $handle:expr,
+        $format:expr,
+        $typ:ty,
+        $ser_func:ident
+    ) => {{
+        let element_ptr = get_obj_pointer!($context, $handle);
+        let element = element_ptr.downcast_ref::<$typ>().unwrap();
+        let mut buf = Vec::new();
+        $context.charge($gas_params.placeholder)?;
+        element.$ser_func(&mut buf).unwrap();
+        buf
+    }};
+}
+
+fn serialize_internal(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    assert_eq!(1, ty_args.len());
+    let handle = safely_pop_arg!(args, u64) as usize;
+    let format_opt = SerializationFormat::try_from(safely_pop_arg!(args, u64));
+    let structure_opt = structure_from_ty_arg!(context, &ty_args[0]);
+    match (structure_opt, format_opt) {
+        (Some(Structure::BLS12381Fr), Ok(SerializationFormat::BLS12381FrLsb)) => {
+            abort_unless_feature_enabled!(context, FeatureFlag::BLS12_381_STRUCTURES);
+            let buf = ark_serialize_internal!(
+                gas_params,
+                context,
+                Structure::BLS12381Fr,
+                handle,
+                SerializationFormat::BLS12381FrLsb,
+                ark_bls12_381::Fr,
+                serialize_uncompressed
```

Comment:
Is `serialzie_uncompressed` the name of an arkworks function? Maybe clarify with a comment.

Reply:
Added an example expansion comment for the macro... Should look better now.